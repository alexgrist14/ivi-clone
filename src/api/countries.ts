import axios from "axios";
import useSWR, { SWRResponse } from "swr";
import { ICountry } from "../types/IMovie";

export const getCountriesSlugs = async (url: string): Promise<string[]> => {
  const response = await axios.get(url);
  const data = response.data as ICountry[];
  const slugs = data
    ? data.map((country: ICountry) => country.slug.toLowerCase())
    : [];

  return slugs;
};

export const useCountriesSlugs = (): SWRResponse<string[]> => {
  return useSWR(
    `${String(process.env.SERVER_HOST)}/countries`,
    getCountriesSlugs
  );
};
