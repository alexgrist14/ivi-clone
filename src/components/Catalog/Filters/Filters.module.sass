.filters
  width: 100%
  background-color: #1f1b2e
  padding: 16px
  border-radius: 6px
  position: relative

  &__top, &__bottom
    display: grid
    column-gap: 12px
    row-gap: 20px
    padding-bottom: 20px

  &__top
    grid-template-columns: repeat(3,  1fr)

  &__bottom
    grid-template-columns: repeat(4, 1fr)

  &__person
    display: grid
    grid-template-columns: 1fr 1fr
    align-items: center
    gap: 10px

  &__buttons
    display: flex
    gap: 20px
    width: 100%
    justify-content: space-between

@media screen and (max-width: 960px)
  .filters
    &__top
      grid-template-columns: repeat(2, 1fr)

      & > div:nth-of-type(3)
        grid-column: 1 / span 2

    &__bottom
      grid-template-columns: repeat(2, 1fr)

@media screen and (max-width: 550px)
  .filters
    &__top, &__bottom
      grid-template-columns: 1fr

      & > div:nth-of-type(3)
        grid-column: auto

    &__buttons
      display: grid
