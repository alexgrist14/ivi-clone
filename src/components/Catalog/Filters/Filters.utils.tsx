import { ReactNode } from "react";
import { BiGhost } from "react-icons/bi";
import { BsJournalRichtext } from "react-icons/bs";
import { FaRedhat } from "react-icons/fa";
import {
  GiBolterGun,
  GiChargedArrow,
  GiDramaMasks,
  GiGhostAlly,
  GiHatchets,
  GiLovers,
  GiMusicalNotes,
  GiWinchesterRifle,
} from "react-icons/gi";
import { MdTheaterComedy } from "react-icons/md";
import { RiTreasureMapLine } from "react-icons/ri";
import { SwiperOptions } from "swiper";
import { TbRating18Plus } from "react-icons/tb";
import { MdSportsBasketball } from "react-icons/md";
import { IFilters } from "/src/types/IFilter";
import { SiWikidotjs } from "react-icons/si";
import { FaChild } from "react-icons/fa";

export const sliderBreakpoints: { [width: number]: SwiperOptions } = {
  0: { slidesPerView: 1, slidesPerGroup: 1 },
  350: { slidesPerView: 2 },
  460: { slidesPerView: 3 },
  590: { slidesPerView: 4 },
  700: { slidesPerView: 5 },
  960: { slidesPerView: 4, slidesPerGroup: 2, spaceBetween: 12 },
};

export const genreIcons: { [key: string]: ReactNode } = {
  drama: <GiDramaMasks size={32} />,
  comedy: <MdTheaterComedy size={32} />,
  action: <GiBolterGun size={32} />,
  thriller: <GiGhostAlly size={32} />,
  western: <GiWinchesterRifle size={32} />,
  adventure: <RiTreasureMapLine size={32} />,
  anime: <GiChargedArrow size={32} />,
  melodrama: <GiLovers size={32} />,
  sport: <MdSportsBasketball size={32} />,
  musical: <GiMusicalNotes size={32} />,
  biography: <BsJournalRichtext size={32} />,
  adults: <TbRating18Plus size={32} />,
  filmnoir: <FaRedhat size={32} />,
  criminal: <GiHatchets size={32} />,
  horror: <BiGhost size={32} />,
  animatedfilm: <SiWikidotjs size={32} />,
  children: <FaChild size={32} />,
};

export const yearFilterItems: string[] = [
  "0-0",
  "2023-2023",
  "2022-2022",
  "2021-2021",
  "2020-2020",
  "2019-2019",
  "2018-2018",
  "2017-2017",
  "2016-2016",
  "2022-2023",
  "2021-2022",
  "2020-2021",
  "2019-2020",
  "2010-2020",
  "2010-2015",
  "2000-2010",
  "1990-2000",
  "1980-1990",
  "0-1990",
];

export const sortHandler = (arr: string[]): string[] =>
  [...arr].sort((a, b) => (a > b ? 1 : -1));

export const getResetfilters = (
  filters: IFilters
): { [key: string]: string | string[] } => {
  return Object.keys(filters).reduce((result, key) => {
    return filters[key].length > 0 ? { ...result, [key]: [] } : result;
  }, {});
};
