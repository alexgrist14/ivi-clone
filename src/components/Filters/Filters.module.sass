.filters
  display: grid
  grid-template-columns: 1fr 1fr 1fr
  width: 100%
  background-color: #1f1b2e
  padding: 16px
  border-radius: 6px
