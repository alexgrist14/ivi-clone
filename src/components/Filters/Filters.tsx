import { FC, useState } from "react";
import { useTranslation } from "react-i18next";
import styles from "./Filters.module.sass";
import {
  countryFilterItems,
  genreFilterItems,
  genreFilterSlides,
  yearFilterItems,
} from "./Filters.utils";
import MultiSelector from "./MultiSelector/MultiSelector";
import PersonSelector from "./PersonSelector/PersonSelector";
import RangeSelector from "./RangeSelector/RangeSelector";
import ResetButton from "./ResetButton/ResetButton";
import SingleSelector from "./SingleSelector/SingleSelector";
import { IFilter } from "/src/types/IFilter";

const sortHandler = (a: IFilter, b: IFilter) => (a.slug > b.slug ? 1 : -1);

const Filters: FC = () => {
  const { t } = useTranslation();

  const [genreFilters, setGenreFilters] = useState<string[]>([]);
  const [countryFilters, setCountryFilters] = useState<string[]>([]);
  const [yearFilter, setYearFilter] = useState<string>("all");
  const [ratingFilter, setRatingFilter] = useState<string>("0");
  const [ratingCountFilter, setRatingCountFilter] = useState<string>("0");
  const [actorFilter, setActorFilter] = useState<string>("");
  const [directorFilter, setDirectorFilter] = useState<string>("");

  return (
    <section className={styles.filters}>
      <MultiSelector
        title={t("filters.genre")}
        items={genreFilterItems.sort(sortHandler)}
        sliderItems={genreFilterSlides}
        activeFilters={genreFilters}
        setActiveFilters={setGenreFilters}
        dropdownPosition="left"
      />
      <MultiSelector
        title={t("filters.country")}
        items={countryFilterItems.sort(sortHandler)}
        sliderItems={[...countryFilterItems].slice(0, 10)}
        activeFilters={countryFilters}
        setActiveFilters={setCountryFilters}
        dropdownPosition="center"
      />
      <SingleSelector
        title={t("filters.years.title")}
        items={yearFilterItems}
        activeFilter={yearFilter}
        setActiveFilter={setYearFilter}
      />
      <RangeSelector
        title={t("filters.rating")}
        max={9}
        step={0.1}
        activeFilter={ratingFilter}
        setActiveFilter={setRatingFilter}
      />
      <RangeSelector
        title={t("filters.ratingCount")}
        max={500000}
        step={5000}
        activeFilter={ratingCountFilter}
        setActiveFilter={setRatingCountFilter}
      />
      <div className={styles.filters__person}>
        <PersonSelector
          type="actor"
          activeFilter={actorFilter}
          setActiveFilter={setActorFilter}
        />
        <PersonSelector
          type="director"
          activeFilter={directorFilter}
          setActiveFilter={setDirectorFilter}
        />
      </div>
      <ResetButton
        clickCallback={() => {
          setGenreFilters([]);
          setCountryFilters([]);
          setYearFilter("all");
          setRatingFilter("0");
          setRatingCountFilter("0");
          setActorFilter("");
          setDirectorFilter("");
        }}
      />
    </section>
  );
};

export default Filters;
