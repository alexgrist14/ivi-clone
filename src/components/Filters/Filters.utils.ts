import {IFilter, IFilterSlide} from "/src/types/IFilter";

export const genreFilterItems: IFilter[] = [
  {slug: 'arthouse', text: 'Артхаус'},
  {slug: 'western', text: 'Вестерн'},
  {slug: 'children', text: 'Для детей'},
  {slug: 'foreign', text: 'Зарубежные'},
  {slug: 'comedy', text: 'Комедии'},
  {slug: 'mystic', text: 'Мистические'},
  {slug: 'adventure', text: 'Приключения'},
  {slug: 'ussr', text: 'Советские'},
  {slug: 'horror', text: 'Ужасы'},
  {slug: 'biography', text: 'Биография'},
  {slug: 'action', text: 'Боевики'},
  {slug: 'military', text: 'Военные'},
] 

export const genreFilterSlides: IFilterSlide[] = [
  {slug: 'arthouse', text: 'Артхаус'},
  {slug: 'western', text: 'Вестерн'},
  {slug: 'children', text: 'Для детей'},
  {slug: 'foreign', text: 'Зарубежные'},
  {slug: 'comedy', text: 'Комедии'},
  {slug: 'mystic', text: 'Мистические'},
]
