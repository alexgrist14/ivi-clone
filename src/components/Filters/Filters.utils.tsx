import {
  GiBolterGun,
  GiDramaMasks,
  GiGhostAlly,
  GiWinchesterRifle,
} from "react-icons/gi";
import { MdTheaterComedy } from "react-icons/md";
import { RiTreasureMapLine } from "react-icons/ri";
import { IFilter, IFilterSlide } from "/src/types/IFilter";

export const genreFilterItems: IFilter[] = [
  { slug: "arthouse", text: "genres.arthouse" },
  { slug: "western", text: "genres.western" },
  { slug: "children", text: "genres.children" },
  { slug: "foreign", text: "genres.foreign" },
  { slug: "comedy", text: "genres.comedy" },
  { slug: "mystic", text: "genres.mystic" },
  { slug: "adventure", text: "genres.adventure" },
  { slug: "ussr", text: "genres.ussr" },
  { slug: "horror", text: "genres.horror" },
  { slug: "biography", text: "genres.biography" },
  { slug: "action", text: "genres.action" },
  { slug: "military", text: "genres.military" },
];

export const genreFilterSlides: IFilterSlide[] = [
  { slug: "drama", text: "genres.drama", icon: <GiDramaMasks size={32} /> },
  {
    slug: "comedy",
    text: "genres.comedy",
    icon: <MdTheaterComedy size={32} />,
  },
  { slug: "action", text: "genres.action", icon: <GiBolterGun size={32} /> },
  {
    slug: "thriller",
    text: "genres.thriller",
    icon: <GiGhostAlly size={32} />,
  },
  {
    slug: "western",
    text: "genres.western",
    icon: <GiWinchesterRifle size={32} />,
  },
  {
    slug: "adventure",
    text: "genres.adventure",
    icon: <RiTreasureMapLine size={32} />,
  },
];

export const countryFilterItems: IFilter[] = [
  { slug: "aus", text: "countries.aus" },
  { slug: "arg", text: "countries.arg" },
  { slug: "arm", text: "countries.arm" },
  { slug: "bel", text: "countries.bel" },
  { slug: "belg", text: "countries.belg" },
  { slug: "bra", text: "countries.bra" },
  { slug: "greatbr", text: "countries.greatbr" },
  { slug: "hung", text: "countries.hung" },
  { slug: "ger", text: "countries.ger" },
  { slug: "hong", text: "countries.hong" },
  { slug: "denm", text: "countries.denm" },
  { slug: "india", text: "countries.india" },
  { slug: "irel", text: "countries.irel" },
  { slug: "spa", text: "countries.spa" },
  { slug: "ita", text: "countries.ita" },
  { slug: "kaz", text: "countries.kaz" },
  { slug: "can", text: "countries.can" },
  { slug: "china", text: "countries.china" },
  { slug: "col", text: "countries.col" },
  { slug: "mex", text: "countries.mex" },
  { slug: "neth", text: "countries.neth" },
  { slug: "newzel", text: "countries.newzel" },
  { slug: "norw", text: "countries.norw" },
  { slug: "pol", text: "countries.pol" },
  { slug: "rus", text: "countries.rus" },
  { slug: "ussr", text: "countries.ussr" },
  { slug: "usa", text: "countries.usa" },
  { slug: "thai", text: "countries.thai" },
  { slug: "turk", text: "countries.turk" },
  { slug: "fin", text: "countries.fin" },
  { slug: "fra", text: "countries.fra" },
  { slug: "switz", text: "countries.switz" },
  { slug: "swed", text: "countries.swed" },
  { slug: "southaf", text: "countries.southaf" },
  { slug: "southkr", text: "countries.southkr" },
  { slug: "jap", text: "countries.jap" },
];

export const yearFilterItems: IFilter[] = [
  { slug: "all", text: "filters.years.all" },
  { slug: "2023", text: "filters.years.2023" },
  { slug: "2022", text: "filters.years.2022" },
  { slug: "2021", text: "filters.years.2021" },
  { slug: "2020", text: "filters.years.2020" },
  { slug: "2019", text: "filters.years.2019" },
  { slug: "2018", text: "filters.years.2018" },
  { slug: "2017", text: "filters.years.2017" },
  { slug: "2016", text: "filters.years.2016" },
  { slug: "2022-2023", text: "" },
  { slug: "2021-2022", text: "" },
  { slug: "2020-2021", text: "" },
  { slug: "2019-2020", text: "" },
  { slug: "2010-2020", text: "" },
  { slug: "2010-2015", text: "" },
  { slug: "2000-2010", text: "" },
  { slug: "1990-2000", text: "" },
  { slug: "1980-1990", text: "" },
  { slug: "0-1990", text: "filters.years.until" },
];
