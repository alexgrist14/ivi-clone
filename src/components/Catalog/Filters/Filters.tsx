import { FC } from "react";
import { useTranslation } from "next-i18next";
import styles from "./Filters.module.sass";
import { yearFilterItems } from "./Filters.utils";
import MultiSelector from "./MultiSelector/MultiSelector";
import PersonSelector from "./PersonSelector/PersonSelector";
import RangeSelector from "./RangeSelector/RangeSelector";
import ResetButton from "./ResetButton/ResetButton";
import SingleSelector from "./SingleSelector/SingleSelector";
import { setQueryParams } from "/src/utils/query";
import { useAppSelector } from "/src/hooks/redux";
import { changeHandler } from "/src/utils/filters/changeHandler";
import { defaultFilters } from "/src/utils/filters/filtersVariables";
import { useRouter } from "next/router";

const sortHandler = (a: string, b: string): 1 | -1 => (a > b ? 1 : -1);

const Filters: FC = () => {
  const { t } = useTranslation("filters");
  const router = useRouter();
  const { filters } = useAppSelector((state) => state.filters);
  const persons = useAppSelector((state) => state.persons);
  const slugs = useAppSelector((state) => state.slugs);

  return (
    <div className={styles.filters}>
      <div className={styles.filters__top}>
        <MultiSelector
          title={t("genre")}
          items={slugs.genresSlugs}
          filters={filters}
          filtersType="genres"
          getFilter={(result) =>
            setQueryParams(router, {
              genres: changeHandler(filters.genres, result),
            })
          }
          dropdownPosition="left"
        />
        <MultiSelector
          title={t("country")}
          items={[...slugs.countriesSlugs].sort(sortHandler)}
          filters={filters}
          filtersType={"countries"}
          getFilter={(result) =>
            setQueryParams(router, {
              countries: changeHandler(filters.countries, result),
            })
          }
          dropdownPosition="center"
        />
        <SingleSelector
          title={t("year")}
          items={yearFilterItems}
          filter={filters.year as string}
          filtersType="year"
          getFilter={(result) => {
            setQueryParams(router, {
              year: changeHandler(filters.year, result, true),
            });
          }}
        />
      </div>
      <div className={styles.filters__bottom}>
        <RangeSelector
          title={t("rating")}
          max={9}
          step={0.1}
          filter={filters.rating as string}
          getFilter={(result) =>
            setQueryParams(router, {
              rating: changeHandler(filters.rating, result, true),
            })
          }
        />
        <RangeSelector
          title={t("ratingCount")}
          max={1000000}
          step={10000}
          filter={filters.assessments as string}
          getFilter={(result) =>
            setQueryParams(router, {
              assessments: changeHandler(filters.assessments, result, true),
            })
          }
        />
        <PersonSelector
          type="actor"
          list={persons.actors}
          filter={filters.actor as string}
          getFilter={(result) =>
            setQueryParams(router, {
              actor: changeHandler(filters.actor, result, true),
            })
          }
        />
        <PersonSelector
          type="director"
          list={persons.directors}
          filter={filters.director as string}
          getFilter={(result) =>
            setQueryParams(router, {
              director: changeHandler(filters.director, result, true),
            })
          }
        />
      </div>
      <ResetButton
        clickCallback={() => setQueryParams(router, defaultFilters)}
      />
    </div>
  );
};

export default Filters;
