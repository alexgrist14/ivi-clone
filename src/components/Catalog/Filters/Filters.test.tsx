import { act } from "@testing-library/react";
import Filters from "./Filters";
import { useGetActors, useGetDirectors } from "/src/api/persons";
import { IProviderRender } from "/src/types/IRender";
import { actorsList } from "/src/utils/mocks/actors";
import { renderWithProviders } from "/src/utils/test-utils";
import mockRouter from "next-router-mock";
import userEvent from "@testing-library/user-event";
import * as utils from "./Filters.utils";
import { countriesListSlugs } from "/src/utils/mocks/countries";
import { setWindowSize } from "/src/store/slices/windowSizeSlice";

jest.mock("next/router", () => require("next-router-mock"));
jest.mock("axios");
jest.mock("src/api/persons");
jest.mock("./Filters.utils", () => ({
  ...jest.requireActual("./Filters.utils"),
  sortHandler: jest.fn(),
}));

describe("Filters component", () => {
  let i = 0;
  let render: IProviderRender;
  let sortHandlerSpy: jest.SpyInstance;

  beforeEach(() => {
    sortHandlerSpy = jest.spyOn(utils, "sortHandler");

    (useGetActors as jest.Mock).mockReturnValue({
      data: i !== 1 ? actorsList : undefined,
    });
    (useGetDirectors as jest.Mock).mockReturnValue({
      data: i !== 1 ? actorsList : undefined,
    });

    (utils.sortHandler as jest.Mock).mockReturnValue(countriesListSlugs);

    render = renderWithProviders(<Filters />);
  });

  afterEach(() => {
    i++;
    jest.resetAllMocks();
  });

  test("component has been rendered correctly", () => {
    const { getByTestId } = render.component;

    expect(getByTestId("filters")).toBeInTheDocument();
  });

  test("render component with undefined actors and directors lists", () => {
    const { getAllByTestId } = render.component;

    expect(getAllByTestId("person-selector").length).toEqual(2);
  });

  test("update filters on query change", async () => {
    await act(() => mockRouter.replace({ query: { genres: "anime" } }));
    expect(render.store.getState().filters.filters.genres).toEqual("anime");
  });

  test("click on filters reset button", async () => {
    const { getAllByTestId } = render.component;
    await act(() => mockRouter.replace({ query: { genres: "anime" } }));
    expect(render.store.getState().filters.filters.genres).toEqual("anime");

    act(() => userEvent.click(getAllByTestId("reset-all-button")[0]));
    expect(render.store.getState().filters.filters.genres).toEqual([]);
  });

  test("click on page reset button", async () => {
    const { getAllByTestId } = render.component;
    await act(() => mockRouter.replace({ query: { page: "2" } }));
    expect(mockRouter.query.page).toEqual("2");

    act(() => userEvent.click(getAllByTestId("reset-all-button")[1]));
    expect(mockRouter.query.page).toEqual([]);
  });

  test("page reset button with window width > 550", () => {
    const { getAllByTestId } = render.component;
    act(() => render.store.dispatch(setWindowSize({ width: 560, height: 1080 })));

    expect(getAllByTestId("reset-all-button")[1]).toMatchSnapshot();
  });

  test("page reset button with window width < 550", () => {
    const { getAllByTestId } = render.component;
    act(() => render.store.dispatch(setWindowSize({ width: 540, height: 1080 })));

    expect(getAllByTestId("reset-all-button")[1]).toMatchSnapshot();
  });

  test("sort handler has been called", () => {
    expect(sortHandlerSpy).toHaveBeenCalled();
  });

  test("sort handler change initial array", async () => {
    jest.unmock("./Filters.utils");

    const utils = await import("./Filters.utils");

    const result = utils.sortHandler(countriesListSlugs);

    expect(result).not.toEqual(countriesListSlugs);
  });
});
