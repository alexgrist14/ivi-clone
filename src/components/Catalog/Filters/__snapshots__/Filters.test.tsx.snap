// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters component page reset button with window width < 550 1`] = `
<button
  class="reset unselectable"
  data-testid="reset-all-button"
>
  <span
    class="reset__icon"
  >
    <svg
      fill="currentColor"
      height="16"
      stroke="currentColor"
      stroke-width="0"
      version="1.1"
      viewBox="0 0 17 17"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g />
      <path
        d="M9.207 8.5l6.646 6.646-0.707 0.707-6.646-6.646-6.646 6.646-0.707-0.707 6.646-6.646-6.647-6.646 0.707-0.707 6.647 6.646 6.646-6.646 0.707 0.707-6.646 6.646z"
      />
    </svg>
  </span>
  <span
    class="reset__text"
  >
    filters:reset.page
  </span>
</button>
`;

exports[`Filters component page reset button with window width > 550 1`] = `
<button
  class="reset unselectable"
  data-testid="reset-all-button"
>
  <span
    class="reset__text"
  >
    filters:reset.page
  </span>
  <span
    class="reset__icon"
  >
    <svg
      fill="currentColor"
      height="16"
      stroke="currentColor"
      stroke-width="0"
      version="1.1"
      viewBox="0 0 17 17"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g />
      <path
        d="M9.207 8.5l6.646 6.646-0.707 0.707-6.646-6.646-6.646 6.646-0.707-0.707 6.646-6.646-6.647-6.646 0.707-0.707 6.647 6.646 6.646-6.646 0.707 0.707-6.646 6.646z"
      />
    </svg>
  </span>
</button>
`;
